### ISSUE TITLE
Tracking Issue for `os_str_display`

### ISSUE THREAD

**Issue by riverbl** [rust-lang/rust#120048]
Feature gate: `#![feature(os_str_display)]`

This is a tracking issue for the `OsStr::display` method.

### Public API

```rust
// std::ffi

pub mod os_str {
    pub struct Display<'a> { /* private fields */ }
}

impl OsStr {
    pub fn display(&self) -> os_str::Display<'_>;
}
```

### Steps / History

- [x] ACP: https://github.com/rust-lang/libs-team/issues/326
- [x] Implementation: #120051
- [x] Final comment period (FCP)[^1]
- [ ] Stabilization PR: #137336

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by ijackson** [comment]
I'm concerned about the name.  I think `OsStr::to_string_lossy` is the better precedent.  `Path` has just `display` and I think this was a mistake.

These methods are really only suitable for use in diagnostics.  Places where the stringlike thing is transformed for further use must necessarily either give up on handling invalid unicode (so should use a fallible method) or do something more complicated and awkward.

**Comment by ijackson** [comment]
(Worst case behaviour from inappropriate use of these methods is modifying the wrong file.)

**Comment by riverbl** [comment]
> I'm concerned about the name. I think `OsStr::to_string_lossy` is the better precedent. `Path` has just `display` and I think this was a mistake.

`Path` has both [`to_string_lossy`](https://doc.rust-lang.org/1.77.1/std/path/struct.Path.html#method.to_string_lossy) and [`display`](https://doc.rust-lang.org/1.77.1/std/path/struct.Path.html#method.display). `OsStr` already has [`to_string_lossy`](https://doc.rust-lang.org/1.77.1/std/ffi/struct.OsStr.html#method.to_string_lossy), which works in the same way as `Path::to_string_lossy`. This issue is for adding [`OsStr::display`](https://doc.rust-lang.org/nightly/std/ffi/struct.OsStr.html#method.display), which would work in the same way as `Path::display`.

> These methods are really only suitable for use in diagnostics. Places where the stringlike thing is transformed for further use must necessarily either give up on handling invalid unicode (so should use a fallible method) or do something more complicated and awkward.

I've encountered situations where `OsStr::display` would be useful for displaying a potentially non Unicode file name to the user ([example](https://github.com/lapce/lapce/blob/9f4c589e9d3ad7cfb018e5d12e449cc1ff3e34dd/lapce-app/src/editor_tab.rs#L288)). In these cases, the lossy Unicode representation of the `OsStr` is not used beyond being displayed.

**Comment by clarfonthey** [comment]
Are there any objections to stabilising this? It's been close to a year.

**Comment by dtolnay** [comment]
@rfcbot fcp merge

**Comment by ijackson** [comment]
Apparently my earlier comment https://github.com/rust-lang/rust/issues/120048#issuecomment-1983581082 was unclear.

I object to stabilising this *with this name*.  It should be called `.display_lossy()`.

That the analogous method on `Path` is called `.display()` is a mistake.   indeed I have seen bugs in real code where `Path::display` was used inappropriately, where the author probably wouldn't have called a method with `lossy` in its name.  That `Path::display` has a name encouraging erroneous use, doesn't mean we should make the same mistake again with `OsStr`.

> I've encountered situations where OsStr::display would be useful for displaying a potentially non Unicode file name to the user ([example](https://github.com/lapce/lapce/blob/9f4c589e9d3ad7cfb018e5d12e449cc1ff3e34dd/lapce-app/src/editor_tab.rs#L288)). In these cases, the lossy Unicode representation of the OsStr is not used beyond being displayed.

For the avoidance of doubt, yes, I agree that this method is useful and ought to exist.

All I mean is that it should have `lossy` in its name.  If people agree with me, I would be happy to make an MR to change the name here.

(I would also love to try to deprecate `Path::display` and replace it with `Path::display_lossy`. Would that be an ACP?)

**Comment by bjoernager** [comment]
I don't think that there is anything inherent that says *display* must equate to *lossless*. The `Display` implementations for `f16`, `f32`, etc. are all lossy in the sense that many raw representations yield the same pretty output. I don't see why `std::os_str::Display` (and `std::path::Display`) can't also discard some data as long as it's done predictably (unless it's the platform-dependence that is the issue). The `display` functions themselves already add a layer of indirection.

<sub>Additionally, I don't see why `Path` and `OsStr` can't just directly implement `Display` as well, given that `Debug` exists.</sub>

**Comment by BurntSushi** [comment]
I agree with @joshtriplett here, and to add, I also put some value on making this consistent in name with `Path::display`.

**Comment by ijackson** [comment]
> Display is not in general expected to be lossless.

Where is this currently stated?  I think it needs to be a lot more prominent.  For example, it's not stated in the trait docs for `Display`.

**Comment by bjoernager** [comment]
> Where is this currently stated? I think it needs to be a lot more prominent. For example, it's not stated in the trait docs for Display.

It's not necessary (or possible) to list every allowed behaviour of a trait. One can usually assume that a specific implementation is allowed as long as the opposite isn't explicitely stated.

**Comment by ijackson** [comment]
> It's not necessary (or possible) to list every allowed behaviour of a trait. One can usually assume that a specific implementation is allowed as long as the opposite isn't explicitely stated.

I'm sorry to say that this is not the attitude I normally expect within the Rust community.  We should be trying to help Rust programmers write correct software.  That means absolutely not accepting "the implementation is allowed to do anything that's not explicitly forbidden".  That way lies C.

I might be persuaded if this were some obscure detail.  But it's not.  It's core to the semantics of the whole trait (and of `ToString::to_string`).  That `.to_string().parse().unwrap()` is buggy is not merely non-obvious; I think it will be surprising to many people.

As I say, trying to use something that came from `Display` (perhaps via `.to_string()`) as the value, elsewhere in the program (or the whole system) is a foreseeable mistake.  One I have come across people actually making.

At the very least we should explain that this is not the intended use of `Display`.  And if we're going for consistency then why do we have `.to_string_lossy()`?

**Comment by joshtriplett** [comment]
> That .to_string().parse().unwrap() is buggy is not merely non-obvious; I think it will be surprising to many people.

> As I say, trying to use something that came from Display (perhaps via .to_string()) as the value, elsewhere in the program (or the whole system) is a foreseeable mistake.

That might make a good clippy lint to add.

**Comment by joshtriplett** [comment]
I submitted PR https://github.com/rust-lang/rust/pull/136687 to expand the documentation and document the established practice that `Display` need not be a full representation of a type, and that `FromStr` need not accept the format that `Display` emits.

**Comment by thaliaarchi** [comment]
FYI, there's currently bugs in the `Display` impl for `OsStr`/`Path`, which I am addressing in https://github.com/rust-lang/rust/pull/136677. It would be nice to get that in at the same release as this stabilizes.
